src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -250,72 +234,105 @@ private static long compilePattern(final byte value) {
     }
 
     /**
-     * A normal Java HashMap does all these safety things like boundary checks... we don't need that, we need speeeed.
-     *
-     * So I've written an extremely simple linear probing hashmap that should work well enough.
+     * Extremely simple linear probing hashmap that should work well enough.
      */
-    class MeasurementRepository {
-        private int tableSize = 1 << 20; // large enough for the contest.
-        private int tableMask = (tableSize - 1);
+    static class MeasurementRepository {
+        private static final int TABLE_SIZE = 1 << 18; // large enough for the contest.
+        private static final int TABLE_MASK = (TABLE_SIZE - 1);
+
+        private final Entry[] table = new Entry[TABLE_SIZE];
+        /**
+         * Separated hashtable, keeps memory local, idea of from franz1981.
+         * And colocate measurements in Entry:
+         */
+        private final int[] hashTable = new int[TABLE_SIZE];
+
+        static final class Entry {
+            private final long[] data;
+            private final String city;
+            private int min, max, count;
+            private long sum;
+
+            Entry(long[] data, String city) {
+                this.data = data;
+                this.city = city;
+                this.min = 1000;
+                this.max = -1000;
+            }
 
-        private MeasurementRepository.Entry[] table = new MeasurementRepository.Entry[tableSize];
+            public void updateWith(int measurement) {
+                min = min(min, measurement);
+                max = max(max, measurement);
+                sum += measurement;
+                count++;
+            }
 
-        record Entry(long address, long[] data, int length, int hash, String city, Measurement measurement) {
+            public Entry mergeWith(Entry entry) {
```

Comment:
Given that entries are scattered in the table, I was fearing this one to not improve performance, while creating smaller snapshots, to be allocated (temporary and spatially) together, would have trade some additional work, with better locality. I could be very wrong given that I didn't measure it, so, how it looks? This change bring some advantage?

Reply:
Didn't change the performance, but eliminates the need to create a new Measurement-record and merge those entries; skips a step.